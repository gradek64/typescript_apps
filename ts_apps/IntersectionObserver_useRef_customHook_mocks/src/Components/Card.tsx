import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Card from '@material-ui/core/Card';
import CardActionArea from '@material-ui/core/CardActionArea';
import CardActions from '@material-ui/core/CardActions';
import CardContent from '@material-ui/core/CardContent';
import CardMedia from '@material-ui/core/CardMedia';
import Button from '@material-ui/core/Button';
import Typography from '@material-ui/core/Typography';

const useStyles = makeStyles({
  root: {
    maxWidth: 345,
    margin: '20px',
  },
});
interface iProps {
  imgURL: string;
  title: string;
  artist: string;
  year: string;
}

export default function ImgMediaCard(props: iProps) {
  const classes = useStyles();
  const { imgURL, title, artist, year } = props;
  return (
    <Card className={classes.root}>
      <CardActionArea>
        <CardMedia
          component="img"
          alt={title}
          height="140"
          image={imgURL}
          title="Contemplative Reptile"
        />
        <CardContent>
          <Typography gutterBottom variant="h5" component="h2">
            {title}
          </Typography>
          <Typography variant="body2" color="textSecondary" component="p">
            {artist}
          </Typography>
          <Typography variant="body2" color="textSecondary" component="p">
            {year}
          </Typography>
        </CardContent>
      </CardActionArea>
      <CardActions>
        <Button size="small" color="primary">
          Share
        </Button>
        <Button size="small" color="primary">
          Learn More
        </Button>
      </CardActions>
    </Card>
  );
}
