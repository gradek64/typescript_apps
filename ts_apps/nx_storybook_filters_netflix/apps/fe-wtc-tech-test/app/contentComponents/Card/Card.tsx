import React, { useState, useCallback, useEffect, useContext } from 'react';
import styled from 'styled-components';
//custom
import RatingStarComp from '../RatingStarComp/Rating';
import IconsStateManagement from './IconsStateManagement/IconsStateManagement';
import { MoviesContext } from '../../../pages/index';
import { calculateAvarageRatings } from './CardUtills/calculateAvarageRating';

const cardColorMapping = {
  default: '#FFF',
  savedNotWatched: '#ED6606',
  watchedAndSaved: '#049452',
  watchedNotSaved: '#EEC907',
};

const CardContainer = styled.div<{
  color: string;
}>`
  display: flex;
  flex-direction: column;
  background-color: ${(props) => props.color || cardColorMapping.default};
  /* align-items: center; */
  max-width: 200px;
  max-height: 300px;
  border-radius: 20px;
  /* justify-content: center; */
  margin: 10px;
  -webkit-box-shadow: -1px 0px 21px -5px rgba(0, 0, 0, 0.53);
  -moz-box-shadow: -1px 0px 21px -5px rgba(0, 0, 0, 0.53);
  box-shadow: -1px 0px 21px -5px rgba(0, 0, 0, 0.53);
`;
const Poster = styled.div`
  border-radius: 10px;
  width: 150px;
  margin: 0 auto;
`;
const PosterImage = styled.div<{
  imgURL: string;
}>`
  width: 150px;
  height: 221px;
  border-radius: 10px;
  background-image: url(${(props) => props.imgURL});
  background-repeat: no-repeat;
  background-size: 100% 100%;
`;
const CardHeader = styled.div`
  display: flex;
  justify-content: space-between;
  width: 100%;
  padding: 10px 10px 0px 10px;
`;
const CardFooter = styled.div`
  width: 100%;
  padding-top: 7px;
`;
interface Props {
  cardId: number;
}
function Card({ cardId }: Props) {
  const [backgroundColor, setBgColor] = useState(cardColorMapping.default);
  const [count, setCount] = useState(0);

  /*
      useCallback(()={},[depends]) and useMemo(()=>{}(),[depends]) 
      and 
      React.memo(ChildComponent) they are somehow related to each other
     
     HOW ?
     useCallBack and useMemo will memoized function so it is unique so it is not being re-created
     with next update of functional component (** explanation below **)

     The child component if also has to stay controlled , so it wont re-render
     unnecessary it needs to have props change check, and this is what React.Memo()
     HOC does. 

      let init = 0
      const callToAction1 = () => {
        init = init + 2
        callToActionCardCount(init)
      };

      if we don t use React.memo() let init = 0 will be called 
      every time component update messing up the counter every time resets to 0. Wrapping 
      with React.memo(component) will prevent unnecessary re-render. Will only re-render if props has changed
      but not otherwise.

      However if we don t care about re-renders and use useState() hook then hook keeps track of 
      value change we count on and re-render doesn t  matter at all for it value.
      Simply useState keeps track of the value we using independently. then React.Memo() is unnecessary

      const [count, setCount] = useState(0);
      const callToAction1 = () => {
        setCount(count + 1)
        callToActionCardCount(count)
      };
  */

  /*
       (** explanation **)
      how javascript creates Objects

     in javascript World if Object has been re-created  (functions are as well objects)
     even with the same name, args etc ...in reality it is not the SAME! 
     because it was re-created again with the same 
     name but with different memory allocations. So the only way to make sure the 
     function is locked is to wrap it with useCallback or useMemo, due to the fact the 
     hook component will re-render everything every time it updates something.
  
  */


  //register setColor once and dont re-render
  // const setColor = useCallback(({ watch, favorite }) => {
  const setColor = ({ watch, favorite }) => {
    switch (true) {
      case !watch && !favorite:
        setBgColor(cardColorMapping.default);
        console.log('cardColorMapping.default')
        break;
      case watch && !favorite:
        setBgColor(cardColorMapping.watchedNotSaved);
        break;
      case !watch && favorite:
        setBgColor(cardColorMapping.savedNotWatched);
        break;
      case watch && favorite:
        setBgColor(cardColorMapping.watchedAndSaved);
        break;
      default:
        break;
    }
  }
  //}, []);

  const increaseCount = useCallback((value) => {
    setCount(value)
  }, [])


  const { movies } = useContext(MoviesContext);
  const { Watched, Saved, Title, Poster: PosterURL, Ratings } = movies[cardId];
  const movieStatus = {
    watch: JSON.parse(Watched.toLowerCase()),
    favorite: JSON.parse(Saved.toLowerCase()),
  };

  useEffect(() => {
    setColor({ ...movieStatus })
  }, [])

  return (
    <CardContainer color={backgroundColor} key={Title}>
      <p>{`count:${count}`}</p>
      <CardHeader>
        <IconsStateManagement
          //callToActionCardColor={setColor}
          callToActionCardCount={increaseCount}
          {...movieStatus}
        />
      </CardHeader>
      <Poster>
        <PosterImage imgURL={PosterURL} />
      </Poster>
      <CardFooter>
        <RatingStarComp rating={calculateAvarageRatings(Ratings)} />
      </CardFooter>
    </CardContainer>
  );
}

export default Card;
