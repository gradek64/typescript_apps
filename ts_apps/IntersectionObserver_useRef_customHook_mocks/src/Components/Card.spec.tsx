import React from 'react';
import { render, act, screen, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import Card from './Card';

const props = {
  year: '2000',
  artist: 'Salvador Dali',
  imgURL: 'https://nrs.harvard.edu/urn-3:HUAM:INV106713N_dynmc',
  title:
    'Untitled (double studio portrait of older woman wearing corsage and top with light front)',
};

afterEach(() => {
  cleanup();
});
describe('renders component', () => {
  it('should render list from dataSet prop', () => {
    const { asFragment } = render(<Card {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
