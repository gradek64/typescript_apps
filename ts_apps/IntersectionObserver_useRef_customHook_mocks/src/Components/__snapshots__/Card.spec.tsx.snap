// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders component should render list from dataSet prop 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiCard-root makeStyles-root-1 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <button
      class="MuiButtonBase-root MuiCardActionArea-root"
      tabindex="0"
      type="button"
    >
      <img
        alt="Untitled (double studio portrait of older woman wearing corsage and top with light front)"
        class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img"
        height="140"
        src="https://nrs.harvard.edu/urn-3:HUAM:INV106713N_dynmc"
        title="Contemplative Reptile"
      />
      <div
        class="MuiCardContent-root"
      >
        <h2
          class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom"
        >
          Untitled (double studio portrait of older woman wearing corsage and top with light front)
        </h2>
        <p
          class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
        >
          Salvador Dali
        </p>
        <p
          class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
        >
          2000
        </p>
      </div>
      <span
        class="MuiCardActionArea-focusHighlight"
      />
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <div
      class="MuiCardActions-root MuiCardActions-spacing"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-textSizeSmall MuiButton-sizeSmall"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Share
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-textSizeSmall MuiButton-sizeSmall"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Learn More
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
